File: Novomber 2022/Day 24/C++ Code/Longest Bitonic subsequence.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution{
    
// 	Time Complexity :- O(n^2);
// 	Space Complexity :- O(n);
	public:
	int LongestBitonicSequence(vector<int>nums)
	{
	    // code here
	    int n = nums.size();
	    vector<int> right(n,0);
        for(int i=n-1;i>=0;i--) {
            for(int j=i+1;j<n;j++) {
                if(nums[i]>nums[j]) right[i] = max(right[i],right[j]);
            }
            right[i]++;
        }
        
        vector<int> left(n,0);
        for(int i=0;i<n;i++) {
            for(int j=0;j<i;j++) {
                if(nums[j]<nums[i]) left[i] = max(left[i],left[j]);
            }
            left[i]++;
        }
        
        int ans=0;
        for(int i=0;i<n;i++) {
            ans = max(ans,left[i]+right[i]-1);
        }
        
        return ans;
	}
};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n;
		cin >> n;
		vector<int>nums(n);
		for(int i = 0; i < n; i++)
			cin >> nums[i];
		Solution ob;
		int ans = ob.LongestBitonicSequence(nums);
		cout << ans <<"\n";
	}
	return 0;
}
// } Driver Code Ends

File: Novomber 2022/Day 24/JAVA Code/Longest Bitonic subsequence.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            int n = Integer.parseInt(br.readLine().trim());
            String s = br.readLine().trim();
            String[] s1 = s.split(" ");
            int[] nums = new int[n];
            for(int i = 0; i < s1.length; i++)
                nums[i] = Integer.parseInt(s1[i]);
            Solution ob = new Solution();
            int ans = ob.LongestBitonicSequence(nums);
            System.out.println(ans);           
        }
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution
{
    public int LongestBitonicSequence(int[] nums)
    {
        // Code here
        int[] right = new int[nums.length];
        for(int i=nums.length-1;i>=0;i--) {
            for(int j=i+1;j<nums.length;j++) {
                if(nums[i]>nums[j]) right[i] = Math.max(right[i],right[j]);
            }
            right[i]++;
        }
        
        int[] left = new int[nums.length];
        for(int i=0;i<nums.length;i++) {
            for(int j=0;j<i;j++) {
                if(nums[j]<nums[i]) left[i] = Math.max(left[i],left[j]);
            }
            left[i]++;
        }
        
        int ans=0;
        for(int i=0;i<left.length;i++) {
            ans = Math.max(ans,left[i]+right[i]-1);
        }
        
        return ans;
    }
}
